index.js

```diff
@@ -92,7 +92,14 @@ function getFirstHeader (req, header) {
 
 function parsePartialURL (url) {
   const containsProtocol = url.indexOf('://') !== -1
-  const result = parseUrl(containsProtocol ? url : 'invalid://' + url)
-  if (!containsProtocol) result.protocol = ''
+  const urlInstance = new URL(containsProtocol ? url : 'invalid://' + url)
+  const result = {}
+  for (const prop in urlInstance) {
```

Comment:
Preferred to use `for..in` loop to not have to maintain a list of URL properties.

Reply:
I haven't run tests, but I believe you can just pass back the `urlInstance` here. No need to copy properties over to another object.

Reply:
Ah, I see that the `URL` instance is resistant to having the `.protocol` and `.href` changed.

```
> result.href = result.href.replace('invalid://', '')
Uncaught TypeError [ERR_INVALID_URL]: Invalid URL
    at __node_internal_captureLargerStackTrace (node:internal/errors:478:5)
    at new NodeError (node:internal/errors:387:5)
    at URL.onParseError (node:internal/url:565:9)
    at URL.set href [as href] (node:internal/url:755:5) {
  input: '//my/path',
  code: 'ERR_INVALID_URL'
}
```

What about pulling the `invalid://` string out to a const INVALID_PROTO, and then in the calling code change this:

```js
if (url.protocol) result.protocol = url.protocol
```

to this:

```js
if (url.protocol !== INVALID_PROTO) result.protocol = url.protocol
```

Then one should be able to avoid copying the URL properties to a new object.
This could be improved upon a little bit to not be susceptible to user code *actually* using "invalid://" as a proto.

Reply:
I just realised other parts of the logic rely on the mutability of the URL object properties to resolve some values [like the protocol](https://github.com/watson/original-url/blob/4447ad715030695b5a9627765e211e018e99c38c/index.js#L47).

So either we complicate the main logic or we keep the property copy. Performance wise I guess 1st option would be the one to pick.

> This could be improved upon a little bit to not be susceptible to user code actually using "invalid://" as a proto.

Do you mean to use a value which is longer and more unlikely to be used by consumers? Something like `original-url-protocol:` or something of the sorts?

Reply:
Did some changes following your advice @trentm 
- changed logic to detect have a placeholder proto and detect it into the main logic. No need to copy properties
- add a try/catch logic to fallback to legacy parse if URL throws

Something I've noticed in the tests is that `parse` method return more properties in the result object like `slashes` and `auth`. IMO this is a breaking change and needs to be reflected in the version with a major bump.